// @mastra/memory/dist/observational-memory-UCOMAMSF.cjs
'use strict';

var chunkQRKB5I2S_cjs = require('./chunk-QRKB5I2S.cjs');



Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_DEFAULTS", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.OBSERVATIONAL_MEMORY_DEFAULTS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_INSTRUCTIONS", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.OBSERVATION_CONTEXT_INSTRUCTIONS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_PROMPT", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.OBSERVATION_CONTEXT_PROMPT; }
});
Object.defineProperty(exports, "OBSERVATION_CONTINUATION_HINT", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.OBSERVATION_CONTINUATION_HINT; }
});
Object.defineProperty(exports, "OBSERVER_SYSTEM_PROMPT", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.OBSERVER_SYSTEM_PROMPT; }
});
Object.defineProperty(exports, "ObservationalMemory", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.ObservationalMemory; }
});
Object.defineProperty(exports, "TokenCounter", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.TokenCounter; }
});
Object.defineProperty(exports, "buildObserverPrompt", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.buildObserverPrompt; }
});
Object.defineProperty(exports, "buildObserverSystemPrompt", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.buildObserverSystemPrompt; }
});
Object.defineProperty(exports, "extractCurrentTask", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.extractCurrentTask; }
});
Object.defineProperty(exports, "formatMessagesForObserver", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.formatMessagesForObserver; }
});
Object.defineProperty(exports, "hasCurrentTaskSection", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.hasCurrentTaskSection; }
});
Object.defineProperty(exports, "optimizeObservationsForContext", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.optimizeObservationsForContext; }
});
Object.defineProperty(exports, "parseObserverOutput", {
  enumerable: true,
  get: function () { return chunkQRKB5I2S_cjs.parseObserverOutput; }
});
